lib/io/QuadSinkHdt.ts

```diff
@@ -1,85 +1,123 @@
-  private readonly poolSize: number;
+  private readonly generateIndexes: boolean;
+  private readonly conversionProgress: boolean;
+  private readonly removeSourceFiles: boolean;
+  private readonly conversionConcurrency: number;
+
+  private readonly docker: Docker;
+  private readonly filesToConvert: Set<string>;
 
-  public constructor(
-    options: IQuadSinkFileOptions,
-    poolSize = 1,
-    deleteSourceFiles = false,
-    errorFileDockerRfdhdt = './error_log_docker_rfdhdt.txt',
-  ) {
+  public constructor(options: IQuadSinkHdtOptions) {
     super(options);
-    this.deleteSourceFiles = deleteSourceFiles;
-    this.errorFileDockerRfdhdt = createWriteStream(errorFileDockerRfdhdt);
-    this.poolSize = poolSize;
+    if (!HDTCPP_FORMATS.has(options.outputFormat)) {
+      throw new Error(`Unsupported HDT output format ${options.outputFormat}`);
+    }
+    this.generateIndexes = options.generateIndexes ?? false;
+    this.conversionProgress = options.conversionProgress ?? false;
+    this.removeSourceFiles = options.removeSourceFiles ?? true;
+    this.conversionConcurrency = options.conversionConcurrency ?? 1;
+    this.docker = new Docker();
+    this.filesToConvert = new Set();
   }
 
-  public async push(iri: string, quad: RDF.Quad): Promise<void> {
-    const path = Path.join('./', this.getFilePath(iri));
-    await super.push(iri, quad);
+  protected getFilePath(iri: string): string {
+    const path = super.getFilePath(iri);
+    this.filesToConvert.add(resolve(path));
+    return path;
+  }
+
+  protected async pullDockerImage(): Promise<void> {
+    await new Promise<void>((resolve, reject) => this.docker.pull(HDTCPP_DOCKER_IMAGE, {}, (error, result) => {
+      if (error) {
+        reject(error);
+      }
+      this.docker.modem.followProgress(result!, error => error ? reject(error) : resolve());
+    }));
+  }
 
-    // Add files with the defined extension to the list to be transformed
-    if (this.fileExtension !== undefined && path.includes(this.fileExtension)) {
-      this.files.add(path);
+  protected async convertSingleFile(rdfFilePath: string): Promise<void> {
+    const cmd = [
+      'rdf2hdt',
+      ...this.generateIndexes ? [ '-i' ] : [],
+      ...this.conversionProgress ? [ '-p' ] : [],
+      `${HDTCPP_MOUNT_PATH}/${basename(rdfFilePath)}`,
+      `${HDTCPP_MOUNT_PATH}/${basename(rdfFilePath).replace(this.fileExtension ?? '', '')}.hdt`,
+    ];
+    const createOptions: Docker.ContainerCreateOptions = {
+      // eslint-disable-next-line ts/naming-convention
+      HostConfig: {
+        // eslint-disable-next-line ts/naming-convention
+        AutoRemove: true,
+        // eslint-disable-next-line ts/naming-convention
+        Binds: [ `${dirname(rdfFilePath)}:${HDTCPP_MOUNT_PATH}:rw` ],
+      },
+    };
+    await this.docker.run(HDTCPP_DOCKER_IMAGE, cmd, process.stdout, createOptions);
+    if (this.removeSourceFiles) {
+      await unlink(rdfFilePath);
```

Comment:
Why unlink since we know for sure they will be files and not symlink?

Reply:
Because the unlink function handles both unlinking symbolic links and deleting of actual files. I just prefer it over the various removal functions, because it works for both, and I also use unlink in Python. :innocent:

Reply:
I don't have a real opinion about it, it just feel a bit "strange" to handle a case that will never happen, but I guess it has no or little overhead and it does not complexify the code.